lib/Middleware.js

```diff
@@ -0,0 +1,219 @@
+  constructor({name}) {
+    super();
+    this.name = name || 'http-signature-strategy';
+    this._helpers = {};
+    this.dereferenceUrlScheme = {did: true, https: true};
+  }
+
+  async authenticate(req) {
+    const self = this;
+
+    // parse the req
+    let parsed;
+    try {
+      parsed = httpSignatureHeader.parseRequest(req);
+    } catch(err) {
+      return self.error(err);
+    }
+
+    const publicKey = {id: parsed.keyId};
+
+    const getKey = self.use('getKey');
+
+    if(!getKey) {
+      return self.error(new HttpSignatureError(
+        'A `getKey` function must be defined.', 'InvalidStateError'));
+    }
+
+    let keyDoc;
+    try {
+      keyDoc = await getKey(publicKey);
+    } catch(err) {
+      return self.error(err);
+    }
+
+    try {
+      await self._verifySignature({keyDoc, parsed});
+    } catch(err) {
+      return self.error(err);
+    }
+
+    const validateRequest = self.use('validateRequest');
+    if(!validateRequest) {
+      return self.error(new HttpSignatureError(
+        'A `validateRequest` function must be defined.', 'InvalidStateError'));
+    }
+    try {
+      await validateRequest(req);
+    } catch(err) {
+      return self.error(err);
+    }
+
+    // TODO: validate nonce?
+
+    const getUser = self.use('getUser') || self.getUser.bind(this);
+
+    // getUser must throw error if the user/identity is not found
+    let user;
+    try {
+      user = await getUser({keyDoc});
+    } catch(err) {
+      return self.error(err);
+    }
+
+    self.success(user);
+  }
+
+  async getUser({keyDoc}) {
+    return {
+      '@context': 'https://w3id.org/identity/v1',
+      id: keyDoc.owner
+    };
+  }
+
+  // decides which scheme to use to look up the public key
+  async ldGetKey({id}) {
+    const self = this;
+    // get scheme from public key ID (URL)
+    let scheme = url.parse(id).protocol || ':';
+    scheme = scheme.substr(0, scheme.length - 1);
+    // dereference URL if supported
+    if(self.dereferenceUrlScheme[scheme]) {
+      if(scheme === 'did') {
+        return self.getDidPublicKey(id);
+      }
+      if(scheme === 'https') {
+        return self.getHttpsPublicKey(id);
+      }
+    }
+    throw new HttpSignatureError(
+      `URL scheme '${scheme}' is not supported.`, 'NotSupportedError');
+  }
+
+  async getDidPublicKey(publicKeyId) {
+    // TODO: utilized did-io API
+    return didio.get(publicKeyId);
+  }
+
+  // 1. resolve key ID => should contain public key info including key material
+  // (and owner)
+  // 2. resolve owner ID and get identity info and list of public keys
+  // 3. make sure key ID is listed in list of keys -- if so, verified
+  async getHttpsPublicKey(publicKeyId) {
+    const self = this;
+    let keyDoc;
+    try {
+      keyDoc = await axios.get(publicKeyId).then(res => res.data);
+    } catch(err) {
+      throw new HttpSignatureError(
+        'Public key URL unavailable.', 'NotFoundError');
+    }
+
+    if(!(keyDoc && typeof keyDoc === 'object')) {
```

Comment:
Should these checks be common in ldGetKey so both https and did key data is checked the same way?  I'm not sure what overlap there will be with did-io.

Reply:
@mattcollier Any thoughts on this?

Reply:
@davidlehn I with you, I don't know what the overlap will be.  I will add a TODO as a reminder to reevaluate this when the new did-io is integrated.